In [5]:
from surprise import SVD, Dataset,SVDpp,NMF
from surprise.model_selection import cross_validate


In [6]:
data = Dataset.load_builtin('ml-100k')

model_SVD = SVD()
model_SVDpp = SVDpp() 
model_NMF = NMF() 


In [7]:
cross_validate(model_SVD,data,measures=['RMSE','MAE'],cv = 5,verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9434  0.9347  0.9379  0.9310  0.9388  0.9372  0.0042  
MAE (testset)     0.7443  0.7359  0.7410  0.7329  0.7402  0.7389  0.0040  
Fit time          0.79    0.76    0.76    0.79    0.84    0.79    0.03    
Test time         0.07    0.10    0.12    0.13    0.07    0.10    0.02    


{'test_rmse': array([0.94343726, 0.93465012, 0.93792997, 0.93102963, 0.9387854 ]),
 'test_mae': array([0.74425737, 0.73585818, 0.74103144, 0.73292206, 0.74020301]),
 'fit_time': (0.789588212966919,
  0.7559177875518799,
  0.7576885223388672,
  0.794640064239502,
  0.84201979637146),
 'test_time': (0.06842803955078125,
  0.10213541984558105,
  0.11661648750305176,
  0.1262495517730713,
  0.06846761703491211)}

In [10]:
cross_validate(model_SVDpp,data,measures=['RMSE','MAE'],cv = 5,verbose=True)

Evaluating RMSE, MAE of algorithm SVDpp on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9199  0.9227  0.9193  0.9268  0.9151  0.9208  0.0039  
MAE (testset)     0.7232  0.7227  0.7197  0.7269  0.7177  0.7220  0.0031  
Fit time          13.20   13.27   13.14   13.09   13.11   13.16   0.07    
Test time         2.50    2.52    2.89    2.52    2.75    2.64    0.16    


{'test_rmse': array([0.9198505 , 0.9227294 , 0.91928247, 0.92679679, 0.91513765]),
 'test_mae': array([0.72319701, 0.72268559, 0.71973867, 0.72688387, 0.71768012]),
 'fit_time': (13.201570987701416,
  13.273795366287231,
  13.136008739471436,
  13.09301495552063,
  13.106884717941284),
 'test_time': (2.5021324157714844,
  2.5188534259796143,
  2.8924999237060547,
  2.5150306224823,
  2.7473442554473877)}

In [9]:
cross_validate(model_NMF,data,measures=['RMSE','MAE'],cv = 5,verbose=True)

Evaluating RMSE, MAE of algorithm NMF on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.9595  0.9636  0.9581  0.9619  0.9625  0.9611  0.0020  
MAE (testset)     0.7530  0.7566  0.7556  0.7560  0.7556  0.7553  0.0012  
Fit time          1.19    1.36    1.13    1.17    1.22    1.21    0.08    
Test time         0.05    0.14    0.07    0.11    0.06    0.09    0.03    


{'test_rmse': array([0.95947022, 0.96358111, 0.95809226, 0.96186121, 0.9625252 ]),
 'test_mae': array([0.7529707 , 0.75661561, 0.75555806, 0.75596346, 0.75555869]),
 'fit_time': (1.193788766860962,
  1.3597614765167236,
  1.127936601638794,
  1.1665799617767334,
  1.223494291305542),
 'test_time': (0.053585052490234375,
  0.13739848136901855,
  0.0656728744506836,
  0.11267971992492676,
  0.06305837631225586)}

Найточніша модель SVD++ - RMSE = 0.9208, MAE = 0.7220 але вона значно довше тренується і передбачає <br>
В SVD RMSE = 0.9372, MAE = 0.7389 трохи гіршою є точність але швидкість навчання і передбачення є значно швидшим <br>
В NMF RMSE = 0.9611, MAE = 0.7553 вища похибка. За часом вона є достатньо швидка<br>

Отже якщо потрібна точність то краще вибирати SVD++, але якщо потрібно щоб був баланс і можна трішки поступитися точністю заради швидкості то я вибирати потрібно SVD модель NMF я взагалі не розглядав би